

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>primer_designer package &mdash; primer-designer 0.0.1 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="primer-designer 0.0.1 documentation" href="index.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> primer-designer
          

          
          </a>

          
            
            
              <div class="version">
                0.0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">Release History</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">primer-designer</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>primer_designer package</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/primer_designer.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="primer-designer-package">
<h1>primer_designer package<a class="headerlink" href="#primer-designer-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-primer_designer.designer">
<span id="primer-designer-designer-module"></span><h2>primer_designer.designer module<a class="headerlink" href="#module-primer_designer.designer" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="primer_designer.designer.PrimerDesigner">
<em class="property">class </em><tt class="descclassname">primer_designer.designer.</tt><tt class="descname">PrimerDesigner</tt><big>(</big><em>folder=None</em>, <em>taxon_for_codon_usage=None</em>, <em>tm='55'</em>, <em>min_amplength='100'</em>, <em>max_amplength='500'</em>, <em>gencode='universal'</em>, <em>mode='primers'</em>, <em>clustype='dna'</em>, <em>amptype='dna_GTR'</em>, <em>email=None</em><big>)</big><a class="reference internal" href="_modules/primer_designer/designer.html#PrimerDesigner"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#primer_designer.designer.PrimerDesigner" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">builtins.object</span></tt></p>
<p>Class for designing primers from FASTA files.</p>
<p>It will send a FASTA alignment to <a class="reference external" href="http://floresta.eead.csic.es/primers4clades/#0">primers4clades</a> in order to design
degenerate primers. Input data needed is an alignment in FASTA format
containing at least 4 sequences.
It is recommended that the beginning of each FASTA sequence description
contains the taxon name between square brackets.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>folder</strong> (<em>str</em>) &#8211; path of folder containing the FASTA file alignments</li>
<li><strong>taxon_for_codon_usage</strong> (<em>str</em>) &#8211; optional taxon name that will be inserted in the
description of FASTA sequences between square
brackets so that can be used by primer4clades
to infer the codon table to use</li>
<li><strong>tm</strong> (<em>str</em>) &#8211; temperature</li>
<li><strong>min_amplength</strong> (<em>str</em>) &#8211; minimum amplicon length</li>
<li><strong>max_amplength</strong> (<em>str</em>) &#8211; maximum amplicon length</li>
<li><strong>gencode</strong> (<em>str</em>) &#8211; genetic code. See below for all available genetic
codes</li>
<li><strong>clustype</strong> (<em>str</em>) &#8211; cluster distance metric: <tt class="docutils literal"><span class="pre">dna</span></tt>, <tt class="docutils literal"><span class="pre">protein</span></tt>.</li>
<li><strong>amptype</strong> (<em>str</em>) &#8211; substitution model used to estimate phylogenetic
information</li>
<li><strong>email</strong> (<em>str</em>) &#8211; your email address so that primer4clades can send
you email with detailed results</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Example</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="c"># The values shown are the default. Change them if needed.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">primer_designer</span> <span class="kn">import</span> <span class="n">PrimerDesigner</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span> <span class="o">=</span> <span class="n">PrimerDesigner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">folder</span> <span class="o">=</span> <span class="s">&quot;alignments&quot;</span>   <span class="c"># folder containing the FASTA file alignments</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">tm</span> <span class="o">=</span> <span class="s">&quot;55&quot;</span>               <span class="c"># annealing temperature</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">min_amplength</span> <span class="o">=</span> <span class="s">&quot;250&quot;</span>   <span class="c"># minimum amplicon length</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">max_amplength</span> <span class="o">=</span> <span class="s">&quot;500&quot;</span>   <span class="c"># maximum amplicon length</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">gencode</span> <span class="o">=</span> <span class="s">&quot;universal&quot;</span>   <span class="c"># see below for all available genetic codes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">mode</span>  <span class="o">=</span> <span class="s">&quot;primers&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">clustype</span> <span class="o">=</span> <span class="s">&quot;dna&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">amptype</span> <span class="o">=</span> <span class="s">&quot;dna_GTRG&quot;</span>    <span class="c"># substitution model used to estimate phylogenetic information</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;youremail@email.com&quot;</span>   <span class="c"># primer4clades will send you an email with very detailed results</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">design_primers</span><span class="p">()</span>
</pre></div>
</div>
<p>The best primer pairs will be printed to your screen. Detailed results will
be saved as HTML files in your alignments folder. But it is recommended if
you also get the results by email. <a class="reference external" href="http://floresta.eead.csic.es/primers4clades/#0">primers4clades</a> will send you one email
for each alignment.
The genetic code table (variable <tt class="docutils literal"><span class="pre">gencode</span></tt>) can be any of the following:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">universal</span></tt> for standard</li>
<li><tt class="docutils literal"><span class="pre">2</span></tt> for vertebrate mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">3</span></tt> for yeast mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">4</span></tt> for mold and protozoa mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">5</span></tt> for invertebrate mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">6</span></tt> for ciliate</li>
<li><tt class="docutils literal"><span class="pre">9</span></tt> for echinoderm and flatworm</li>
<li><tt class="docutils literal"><span class="pre">10</span></tt> for  euplotid nuclear</li>
<li><tt class="docutils literal"><span class="pre">11</span></tt> for  bacterial and plastid</li>
<li><tt class="docutils literal"><span class="pre">12</span></tt> for  alternative yeast nuclear</li>
<li><tt class="docutils literal"><span class="pre">13</span></tt> for  ascidian mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">14</span></tt> for  flatworm mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">15</span></tt> for  Blepharisma nuclear</li>
<li><tt class="docutils literal"><span class="pre">16</span></tt> for  Chlorophycean mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">21</span></tt> for  Trematode mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">22</span></tt> for  Scenedesmus obliquus mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">23</span></tt> for  Thraustochytrium mitochondrial</li>
</ul>
</div></blockquote>
<p>The evolutionary substitution model can be any of the following (variable
<tt class="docutils literal"><span class="pre">amptype</span></tt>):</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">protein_WAGG</span></tt>  for protein WAG+G</li>
<li><tt class="docutils literal"><span class="pre">protein_JTTG</span></tt>  for protein JTT+G</li>
<li><tt class="docutils literal"><span class="pre">protein_Blosum62G</span></tt>  for protein Blosum62+G</li>
<li><tt class="docutils literal"><span class="pre">protein_VTG</span></tt>  for protein VT+G</li>
<li><tt class="docutils literal"><span class="pre">protein_DayhoffG</span></tt>  for protein Dayhoff+G</li>
<li><tt class="docutils literal"><span class="pre">protein_MtREVG</span></tt>  for protein MtREV+G</li>
<li><tt class="docutils literal"><span class="pre">dna_HKYG</span></tt>  for dna HKY+G</li>
<li><tt class="docutils literal"><span class="pre">dna_GTRG</span></tt>  for dna GTR+G</li>
<li><tt class="docutils literal"><span class="pre">dna_K80G</span></tt>  for dna K80+G</li>
<li><tt class="docutils literal"><span class="pre">dna_TrNG</span></tt>  for dna TrN+G</li>
<li><tt class="docutils literal"><span class="pre">dna_JC69G</span></tt>  for dna JC69+G</li>
</ul>
</div></blockquote>
<dl class="method">
<dt id="primer_designer.designer.PrimerDesigner.design_primers">
<tt class="descname">design_primers</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/primer_designer/designer.html#PrimerDesigner.design_primers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#primer_designer.designer.PrimerDesigner.design_primers" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="primer_designer.designer.PrimerDesigner.insert_taxon_in_new_fasta_file">
<tt class="descname">insert_taxon_in_new_fasta_file</tt><big>(</big><em>aln</em><big>)</big><a class="reference internal" href="_modules/primer_designer/designer.html#PrimerDesigner.insert_taxon_in_new_fasta_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#primer_designer.designer.PrimerDesigner.insert_taxon_in_new_fasta_file" title="Permalink to this definition">¶</a></dt>
<dd><p>primer4clades infers the codon usage table from the taxon names in the
sequences.</p>
<p>These names need to be enclosed by square brackets and be
present in the description of the FASTA sequence. The position is not
important. I will insert the names in the description in a new FASTA
file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">Filename of modified FASTA file that includes the name of the taxon.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="primer_designer.designer.PrimerDesigner.process_response">
<tt class="descname">process_response</tt><big>(</big><em>aln</em>, <em>response_body</em><big>)</big><a class="reference internal" href="_modules/primer_designer/designer.html#PrimerDesigner.process_response"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#primer_designer.designer.PrimerDesigner.process_response" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="primer_designer.designer.PrimerDesigner.request_primers">
<tt class="descname">request_primers</tt><big>(</big><em>aln</em><big>)</big><a class="reference internal" href="_modules/primer_designer/designer.html#PrimerDesigner.request_primers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#primer_designer.designer.PrimerDesigner.request_primers" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-primer_designer.utils">
<span id="primer-designer-utils-module"></span><h2>primer_designer.utils module<a class="headerlink" href="#module-primer_designer.utils" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="primer_designer.utils.is_fasta">
<tt class="descclassname">primer_designer.utils.</tt><tt class="descname">is_fasta</tt><big>(</big><em>filename</em><big>)</big><a class="reference internal" href="_modules/primer_designer/utils.html#is_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#primer_designer.utils.is_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if filename is FASTA based on extension</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">Boolean</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-primer_designer">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-primer_designer" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="primer_designer.PrimerDesigner">
<em class="property">class </em><tt class="descclassname">primer_designer.</tt><tt class="descname">PrimerDesigner</tt><big>(</big><em>folder=None</em>, <em>taxon_for_codon_usage=None</em>, <em>tm='55'</em>, <em>min_amplength='100'</em>, <em>max_amplength='500'</em>, <em>gencode='universal'</em>, <em>mode='primers'</em>, <em>clustype='dna'</em>, <em>amptype='dna_GTR'</em>, <em>email=None</em><big>)</big><a class="headerlink" href="#primer_designer.PrimerDesigner" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">builtins.object</span></tt></p>
<p>Class for designing primers from FASTA files.</p>
<p>It will send a FASTA alignment to <a class="reference external" href="http://floresta.eead.csic.es/primers4clades/#0">primers4clades</a> in order to design
degenerate primers. Input data needed is an alignment in FASTA format
containing at least 4 sequences.
It is recommended that the beginning of each FASTA sequence description
contains the taxon name between square brackets.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>folder</strong> (<em>str</em>) &#8211; path of folder containing the FASTA file alignments</li>
<li><strong>taxon_for_codon_usage</strong> (<em>str</em>) &#8211; optional taxon name that will be inserted in the
description of FASTA sequences between square
brackets so that can be used by primer4clades
to infer the codon table to use</li>
<li><strong>tm</strong> (<em>str</em>) &#8211; temperature</li>
<li><strong>min_amplength</strong> (<em>str</em>) &#8211; minimum amplicon length</li>
<li><strong>max_amplength</strong> (<em>str</em>) &#8211; maximum amplicon length</li>
<li><strong>gencode</strong> (<em>str</em>) &#8211; genetic code. See below for all available genetic
codes</li>
<li><strong>clustype</strong> (<em>str</em>) &#8211; cluster distance metric: <tt class="docutils literal"><span class="pre">dna</span></tt>, <tt class="docutils literal"><span class="pre">protein</span></tt>.</li>
<li><strong>amptype</strong> (<em>str</em>) &#8211; substitution model used to estimate phylogenetic
information</li>
<li><strong>email</strong> (<em>str</em>) &#8211; your email address so that primer4clades can send
you email with detailed results</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Example</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="c"># The values shown are the default. Change them if needed.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">primer_designer</span> <span class="kn">import</span> <span class="n">PrimerDesigner</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span> <span class="o">=</span> <span class="n">PrimerDesigner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">folder</span> <span class="o">=</span> <span class="s">&quot;alignments&quot;</span>   <span class="c"># folder containing the FASTA file alignments</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">tm</span> <span class="o">=</span> <span class="s">&quot;55&quot;</span>               <span class="c"># annealing temperature</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">min_amplength</span> <span class="o">=</span> <span class="s">&quot;250&quot;</span>   <span class="c"># minimum amplicon length</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">max_amplength</span> <span class="o">=</span> <span class="s">&quot;500&quot;</span>   <span class="c"># maximum amplicon length</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">gencode</span> <span class="o">=</span> <span class="s">&quot;universal&quot;</span>   <span class="c"># see below for all available genetic codes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">mode</span>  <span class="o">=</span> <span class="s">&quot;primers&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">clustype</span> <span class="o">=</span> <span class="s">&quot;dna&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">amptype</span> <span class="o">=</span> <span class="s">&quot;dna_GTRG&quot;</span>    <span class="c"># substitution model used to estimate phylogenetic information</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;youremail@email.com&quot;</span>   <span class="c"># primer4clades will send you an email with very detailed results</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">design_primers</span><span class="p">()</span>
</pre></div>
</div>
<p>The best primer pairs will be printed to your screen. Detailed results will
be saved as HTML files in your alignments folder. But it is recommended if
you also get the results by email. <a class="reference external" href="http://floresta.eead.csic.es/primers4clades/#0">primers4clades</a> will send you one email
for each alignment.
The genetic code table (variable <tt class="docutils literal"><span class="pre">gencode</span></tt>) can be any of the following:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">universal</span></tt> for standard</li>
<li><tt class="docutils literal"><span class="pre">2</span></tt> for vertebrate mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">3</span></tt> for yeast mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">4</span></tt> for mold and protozoa mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">5</span></tt> for invertebrate mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">6</span></tt> for ciliate</li>
<li><tt class="docutils literal"><span class="pre">9</span></tt> for echinoderm and flatworm</li>
<li><tt class="docutils literal"><span class="pre">10</span></tt> for  euplotid nuclear</li>
<li><tt class="docutils literal"><span class="pre">11</span></tt> for  bacterial and plastid</li>
<li><tt class="docutils literal"><span class="pre">12</span></tt> for  alternative yeast nuclear</li>
<li><tt class="docutils literal"><span class="pre">13</span></tt> for  ascidian mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">14</span></tt> for  flatworm mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">15</span></tt> for  Blepharisma nuclear</li>
<li><tt class="docutils literal"><span class="pre">16</span></tt> for  Chlorophycean mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">21</span></tt> for  Trematode mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">22</span></tt> for  Scenedesmus obliquus mitochondrial</li>
<li><tt class="docutils literal"><span class="pre">23</span></tt> for  Thraustochytrium mitochondrial</li>
</ul>
</div></blockquote>
<p>The evolutionary substitution model can be any of the following (variable
<tt class="docutils literal"><span class="pre">amptype</span></tt>):</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">protein_WAGG</span></tt>  for protein WAG+G</li>
<li><tt class="docutils literal"><span class="pre">protein_JTTG</span></tt>  for protein JTT+G</li>
<li><tt class="docutils literal"><span class="pre">protein_Blosum62G</span></tt>  for protein Blosum62+G</li>
<li><tt class="docutils literal"><span class="pre">protein_VTG</span></tt>  for protein VT+G</li>
<li><tt class="docutils literal"><span class="pre">protein_DayhoffG</span></tt>  for protein Dayhoff+G</li>
<li><tt class="docutils literal"><span class="pre">protein_MtREVG</span></tt>  for protein MtREV+G</li>
<li><tt class="docutils literal"><span class="pre">dna_HKYG</span></tt>  for dna HKY+G</li>
<li><tt class="docutils literal"><span class="pre">dna_GTRG</span></tt>  for dna GTR+G</li>
<li><tt class="docutils literal"><span class="pre">dna_K80G</span></tt>  for dna K80+G</li>
<li><tt class="docutils literal"><span class="pre">dna_TrNG</span></tt>  for dna TrN+G</li>
<li><tt class="docutils literal"><span class="pre">dna_JC69G</span></tt>  for dna JC69+G</li>
</ul>
</div></blockquote>
<dl class="method">
<dt id="primer_designer.PrimerDesigner.design_primers">
<tt class="descname">design_primers</tt><big>(</big><big>)</big><a class="headerlink" href="#primer_designer.PrimerDesigner.design_primers" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="primer_designer.PrimerDesigner.insert_taxon_in_new_fasta_file">
<tt class="descname">insert_taxon_in_new_fasta_file</tt><big>(</big><em>aln</em><big>)</big><a class="headerlink" href="#primer_designer.PrimerDesigner.insert_taxon_in_new_fasta_file" title="Permalink to this definition">¶</a></dt>
<dd><p>primer4clades infers the codon usage table from the taxon names in the
sequences.</p>
<p>These names need to be enclosed by square brackets and be
present in the description of the FASTA sequence. The position is not
important. I will insert the names in the description in a new FASTA
file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">Filename of modified FASTA file that includes the name of the taxon.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="primer_designer.PrimerDesigner.process_response">
<tt class="descname">process_response</tt><big>(</big><em>aln</em>, <em>response_body</em><big>)</big><a class="headerlink" href="#primer_designer.PrimerDesigner.process_response" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="primer_designer.PrimerDesigner.request_primers">
<tt class="descname">request_primers</tt><big>(</big><em>aln</em><big>)</big><a class="headerlink" href="#primer_designer.PrimerDesigner.request_primers" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
</div>


           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015, Carlos Peña.
      Last updated on Oct 22, 2015.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.0.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>