<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>PhyloXML &mdash; A Python Environment for (phylogenetic) Tree Exploration</title>
    
    <link rel="stylesheet" href="../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '2.3.10',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="top" title="A Python Environment for (phylogenetic) Tree Exploration" href="../index.html" />
    <link rel="up" title="Phylogenetic XML standards" href="tutorial_xml.html" />
    <link rel="next" title="Interactive web tree visualization" href="tutorial_webplugin.html" />
    <link rel="prev" title="NeXML" href="tutorial_nexml.html" />
<script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
           (i[r].q=i[r].q||[]).push(arguments)
    },i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

 ga('create', 'UA-783841-4', 'etetoolkit.org');
 ga('send', 'pageview');
 </script>

<!-- 
 <script type="text/javascript">
 var _gaq = _gaq || [];
 _gaq.push(['_setAccount', 'UA-783841-12']);
 _gaq.push(['_setCookiePath', '/ete2/']);
 _gaq.push(['_trackPageview']);

 (function() {
 var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
 ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
 var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
 })();
 </script>
 -->
<!-- <link rel="canonical" href="http://etetoolkit.org/docs/2.3/CANONICALPATH" />
<meta http-equiv="refresh" content="0; url=http:/etetoolkit.org/CANONICALPATH" />
<script type="text/javascript">
window.location.replace('http://etetoolkit.org/docs/2.3/' + window.location.pathname.replace('/ete2/',''));
</script> -->
 
 

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_webplugin.html" title="Interactive web tree visualization"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_nexml.html" title="NeXML"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >The ETE tutorial</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="tutorial_xml.html" accesskey="U">Phylogenetic XML standards</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="versionadded">
<p><span class="versionmodified">New in version 2.1.</span></p>
</div>
<div class="section" id="phyloxml">
<h1>PhyloXML<a class="headerlink" href="#phyloxml" title="Permalink to this headline">¶</a></h1>
<p>PhyloXML (<a class="reference external" href="http://www.phyloxml.org/">http://www.phyloxml.org/</a>) is a novel standard used to encode
phylogenetic information. In particular, phyloXML is designed to
describe phylogenetic trees (or networks) and associated data, such as
taxonomic information, gene names and identifiers, branch lengths,
support values, and gene duplication and speciation events.</p>
<div class="section" id="loading-phyloxml-projects-from-files">
<h2>Loading PhyloXML projects from files<a class="headerlink" href="#loading-phyloxml-projects-from-files" title="Permalink to this headline">¶</a></h2>
<p>ETE provides full support for phyloXML projects through the
<a class="reference internal" href="../reference/reference_phyloxml.html#ete2.Phyloxml" title="ete2.Phyloxml"><code class="xref py py-class docutils literal"><span class="pre">Phyloxml</span></code></a> object. Phylogenies are integrated as ETE&#8217;s tree
data structures as <a class="reference internal" href="../reference/reference_phyloxml.html#ete2.PhyloxmlTree" title="ete2.PhyloxmlTree"><code class="xref py py-class docutils literal"><span class="pre">PhyloxmlTree</span></code></a> instances, while the rest of
features are represented as simple classes (<a class="reference internal" href="../reference/reference_phyloxml.html#module-ete2.phyloxml" title="ete2.phyloxml"><code class="xref py py-mod docutils literal"><span class="pre">ete2.phyloxml</span></code></a>)
providing basic reading and writing operations.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete2</span> <span class="kn">import</span> <span class="n">Phyloxml</span>
<span class="n">project</span> <span class="o">=</span> <span class="n">Phyloxml</span><span class="p">()</span>
<span class="n">project</span><span class="o">.</span><span class="n">build_from_file</span><span class="p">(</span><span class="s">&quot;apaf.xml&quot;</span><span class="p">)</span>

<span class="c"># Each tree contains the same methods as a PhyloTree object</span>
<span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">project</span><span class="o">.</span><span class="n">get_phylogeny</span><span class="p">():</span>
    <span class="k">print</span> <span class="n">tree</span>
    <span class="c"># you can even use rendering options</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="c"># PhyloXML features are stored in the phyloxml_clade attribute</span>
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="p">:</span> 
        <span class="k">print</span> <span class="s">&quot;Node name:&quot;</span><span class="p">,</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span>
        <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">phyloxml_clade</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">():</span> 
            <span class="k">for</span> <span class="n">domain</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">domain_architecture</span><span class="o">.</span><span class="n">get_domain</span><span class="p">():</span>
                <span class="n">domain_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">domain</span><span class="o">.</span><span class="n">valueOf_</span><span class="p">,</span> <span class="n">domain</span><span class="o">.</span><span class="n">get_from</span><span class="p">(),</span> <span class="n">domain</span><span class="o">.</span><span class="n">get_to</span><span class="p">()]</span>
                <span class="k">print</span> <span class="s">&quot;  Domain:&quot;</span><span class="p">,</span> <span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">domain_data</span><span class="p">))</span>
</pre></div>
</div>
<p><a class="reference download internal" href="../_downloads/phyloxml_parser.py"><code class="xref download docutils literal"><span class="pre">[Download</span> <span class="pre">script]</span></code></a>
<a class="reference download internal" href="../_downloads/apaf.xml"><code class="xref download docutils literal"><span class="pre">[Download</span> <span class="pre">example]</span></code></a></p>
<p>Each tree node contains two phyloxml elements, <code class="xref py py-attr docutils literal"><span class="pre">phyloxml_clade</span></code>
and <code class="xref py py-attr docutils literal"><span class="pre">phyloxml_phylogeny</span></code>. The first attribute contains clade
information referred to the node, while <code class="docutils literal"><span class="pre">phyloxml_phylogeny</span></code>
contains general data about the subtree defined by each node. This
way, you can split, or copy any part of a tree and it will be exported
as a separate phyloxml phylogeny instance.</p>
<p>Note that <code class="xref py py-attr docutils literal"><span class="pre">node.dist</span></code>, <code class="xref py py-attr docutils literal"><span class="pre">node.support</span></code> and
<code class="xref py py-attr docutils literal"><span class="pre">node.name</span></code> features are linked to
<code class="xref py py-attr docutils literal"><span class="pre">node.phyloxml_clade.branch_length</span></code>,
<code class="xref py py-attr docutils literal"><span class="pre">node.phyloxml_clade.confidence</span></code> and
<code class="xref py py-attr docutils literal"><span class="pre">node.phyloxml_clade.name</span></code>, respectively.</p>
</div>
<div class="section" id="creating-phyloxml-projects-from-scratch">
<h2>Creating PhyloXML projects from scratch<a class="headerlink" href="#creating-phyloxml-projects-from-scratch" title="Permalink to this headline">¶</a></h2>
<p>In order to create new PhyloXML projects, a set of classes is
available in the <a class="reference internal" href="../reference/reference_phyloxml.html#module-ete2.phyloxml" title="ete2.phyloxml"><code class="xref py py-mod docutils literal"><span class="pre">ete2.phyloxml</span></code></a> module.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete2</span> <span class="kn">import</span> <span class="n">Phyloxml</span><span class="p">,</span> <span class="n">phyloxml</span>
<span class="kn">import</span> <span class="nn">random</span> 
<span class="n">project</span> <span class="o">=</span> <span class="n">Phyloxml</span><span class="p">()</span>

<span class="c"># Creates a random tree</span>
<span class="n">phylo</span> <span class="o">=</span> <span class="n">phyloxml</span><span class="o">.</span><span class="n">PhyloxmlTree</span><span class="p">()</span>
<span class="n">phylo</span><span class="o">.</span><span class="n">populate</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">random_branches</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">phylo</span><span class="o">.</span><span class="n">phyloxml_phylogeny</span><span class="o">.</span><span class="n">set_name</span><span class="p">(</span><span class="s">&quot;test_tree&quot;</span><span class="p">)</span>
<span class="c"># Add the tree to the phyloxml project</span>
<span class="n">project</span><span class="o">.</span><span class="n">add_phylogeny</span><span class="p">(</span><span class="n">phylo</span><span class="p">)</span>

<span class="k">print</span> <span class="n">project</span><span class="o">.</span><span class="n">get_phylogeny</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>

<span class="c">#          /-iajom</span>
<span class="c">#     /---|</span>
<span class="c">#    |     \-wiszh</span>
<span class="c">#----|</span>
<span class="c">#    |     /-xrygw</span>
<span class="c">#     \---|</span>
<span class="c">#         |     /-gjlwx</span>
<span class="c">#          \---|</span>
<span class="c">#               \-ijvnk</span>

<span class="c"># Trees can be operated as normal ETE trees</span>
<span class="n">phylo</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>


<span class="c"># Export the project as phyloXML format</span>
<span class="n">project</span><span class="o">.</span><span class="n">export</span><span class="p">()</span>

<span class="c"># &lt;phy:Phyloxml xmlns:phy=&quot;http://www.phyloxml.org/1.10/phyloxml.xsd&quot;&gt;</span>
<span class="c">#     &lt;phy:phylogeny&gt;</span>
<span class="c">#         &lt;phy:name&gt;test_tree&lt;/phy:name&gt;</span>
<span class="c">#         &lt;phy:clade&gt;</span>
<span class="c">#             &lt;phy:name&gt;NoName&lt;/phy:name&gt;</span>
<span class="c">#             &lt;phy:branch_length&gt;0.000000e+00&lt;/phy:branch_length&gt;</span>
<span class="c">#             &lt;phy:confidence type=&quot;branch_support&quot;&gt;1.0&lt;/phy:confidence&gt;</span>
<span class="c">#             &lt;phy:clade&gt;</span>
<span class="c">#                 &lt;phy:name&gt;NoName&lt;/phy:name&gt;</span>
<span class="c">#                 &lt;phy:branch_length&gt;1.665083e-01&lt;/phy:branch_length&gt;</span>
<span class="c">#                 &lt;phy:confidence type=&quot;branch_support&quot;&gt;0.938507980435&lt;/phy:confidence&gt;</span>
<span class="c">#                 &lt;phy:clade&gt;</span>
<span class="c">#                     &lt;phy:name&gt;NoName&lt;/phy:name&gt;</span>
<span class="c">#                     &lt;phy:branch_length&gt;1.366655e-01&lt;/phy:branch_length&gt;</span>
<span class="c">#                     &lt;phy:confidence type=&quot;branch_support&quot;&gt;0.791888248212&lt;/phy:confidence&gt;</span>
<span class="c">#                     &lt;phy:clade&gt;</span>
<span class="c">#                         &lt;phy:name&gt;ojnfg&lt;/phy:name&gt;</span>
<span class="c">#                         &lt;phy:branch_length&gt;2.194209e-01&lt;/phy:branch_length&gt;</span>
<span class="c">#                         &lt;phy:confidence type=&quot;branch_support&quot;&gt;0.304705977822&lt;/phy:confidence&gt;</span>
<span class="c">#                     &lt;/phy:clade&gt;</span>
<span class="c">#                     &lt;phy:clade&gt;</span>
<span class="c">#                         &lt;phy:name&gt;qrfnz&lt;/phy:name&gt;</span>
<span class="c">#                         &lt;phy:branch_length&gt;5.235437e-02&lt;/phy:branch_length&gt;</span>
<span class="c">#                         &lt;phy:confidence type=&quot;branch_support&quot;&gt;0.508533765418&lt;/phy:confidence&gt;</span>
<span class="c">#                     &lt;/phy:clade&gt;</span>
<span class="c">#                 &lt;/phy:clade&gt;</span>
<span class="c">#                 &lt;phy:clade&gt;</span>
<span class="c">#                     &lt;phy:name&gt;shngq&lt;/phy:name&gt;</span>
<span class="c">#                     &lt;phy:branch_length&gt;9.740958e-01&lt;/phy:branch_length&gt;</span>
<span class="c">#                     &lt;phy:confidence type=&quot;branch_support&quot;&gt;0.642187390965&lt;/phy:confidence&gt;</span>
<span class="c">#                 &lt;/phy:clade&gt;</span>
<span class="c">#             &lt;/phy:clade&gt;</span>
<span class="c">#             &lt;phy:clade&gt;</span>
<span class="c">#                 &lt;phy:name&gt;NoName&lt;/phy:name&gt;</span>
<span class="c">#                 &lt;phy:branch_length&gt;3.806412e-01&lt;/phy:branch_length&gt;</span>
<span class="c">#                 &lt;phy:confidence type=&quot;branch_support&quot;&gt;0.383619811911&lt;/phy:confidence&gt;</span>
<span class="c">#                 &lt;phy:clade&gt;</span>
<span class="c">#                     &lt;phy:name&gt;vfmnk&lt;/phy:name&gt;</span>
<span class="c">#                     &lt;phy:branch_length&gt;6.495163e-01&lt;/phy:branch_length&gt;</span>
<span class="c">#                     &lt;phy:confidence type=&quot;branch_support&quot;&gt;0.141298879514&lt;/phy:confidence&gt;</span>
<span class="c">#                 &lt;/phy:clade&gt;</span>
<span class="c">#                 &lt;phy:clade&gt;</span>
<span class="c">#                     &lt;phy:name&gt;btexi&lt;/phy:name&gt;</span>
<span class="c">#                     &lt;phy:branch_length&gt;5.704955e-01&lt;/phy:branch_length&gt;</span>
<span class="c">#                     &lt;phy:confidence type=&quot;branch_support&quot;&gt;0.951876078012&lt;/phy:confidence&gt;</span>
<span class="c">#                 &lt;/phy:clade&gt;</span>
<span class="c">#             &lt;/phy:clade&gt;</span>
<span class="c">#         &lt;/phy:clade&gt;</span>
<span class="c">#     &lt;/phy:phylogeny&gt;</span>
<span class="c"># &lt;/phy:Phyloxml&gt;</span>
</pre></div>
</div>
<p><a class="reference download internal" href="../_downloads/phyloxml_from_scratch.py"><code class="xref download docutils literal"><span class="pre">[Download</span> <span class="pre">script]</span></code></a></p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/ete_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">PhyloXML</a><ul>
<li><a class="reference internal" href="#loading-phyloxml-projects-from-files">Loading PhyloXML projects from files</a></li>
<li><a class="reference internal" href="#creating-phyloxml-projects-from-scratch">Creating PhyloXML projects from scratch</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_nexml.html"
                        title="previous chapter">NeXML</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_webplugin.html"
                        title="next chapter">Interactive web tree visualization</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/tutorial/tutorial_phyloxml.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_webplugin.html" title="Interactive web tree visualization"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_nexml.html" title="NeXML"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >The ETE tutorial</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="tutorial_xml.html" >Phylogenetic XML standards</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Jaime Huerta-Cepas.
      Last updated on Nov 05, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>