<!DOCTYPE html
	PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	 "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
<title>primers4clades job 26e8c759</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
</head>
<body>
<br><h4>primers4clades job  (26e8c759)
</h4><pre>_runmode : primers
_gencode : universal
_evaluation : 
_cluster distance metric : protein
_Tm : 55
_amplicon_length : 200 , 500
_email : youremail@email.com
_FASTA sequence file : Ca2.fst
</pre>
<h4>Clustering sequences...</h4><pre># number of sequences read = 8
# number of recognised taxa = 1
# input sequences <strong>seem to be aligned</strong>
# computing distance matrix...
# multiple alignment <a href="http://floresta.eead.csic.es/primers4clades/showjobfile.cgi?26e8c759_intronless.faa" target="_new">FASTA file</a>
# alignment stats: length = 510 %gaps = 0.00 %constant = 92.0

# mean distance = 0.03258
# max distance = 0.06065 ( 001 <=> 005 )
# min distance = 0.01013 ( 002 <=> 008 )

# NJ tree with cluster labels :


      +004__<b>0</b> [Bombyx_mori]_Seq4_...       
    +-1 
  +-2 +005__<b>0</b> [Bombyx_mori]_Seq5_...       
  ! ! 
  ! +006__<b>0</b> [Bombyx_mori]_Seq6_...       
  ! 
  ! +003__<b>0</b> [Bombyx_mori]_Seq3_...       
  3-4 
  ! ! +007__<b>0</b> [Bombyx_mori]_Seq7_...       
  ! +-5 
  !   ! +002__<b>0</b> [Bombyx_mori]_Seq2_...       
  !   +-6 
  !     +008__<b>0</b> [Bombyx_mori]_Seq8_...       
  ! 
  +-001__<b>0</b> [Bombyx_mori]_Seq1_...       

</pre>
<h4>Calculating primers, results will be shortly emailed...</h4><h4>You might <a href="http://floresta.eead.csic.es/primers4clades//">return</a> or wait for your results to appear here.</h4><pre>
# largest cluster (0) selected, please use runmode 'cluster sequences' to make your own selection

# input sequences seem to be aligned
## table redundancy vs amplicons stats:
# input_derived_4e6be5b3 redundancy = 0.93 amplicons = 4
# Bombyx_mori redundancy = 0.93 amplicons = 4

# evaluating primers...
# primer evaluation <a href="http://floresta.eead.csic.es/primers4clades/showjobfile.cgi?26e8c759_intronless_cluster_0.fna_primers.list.oligo_eval.tab" target="_new" title="tab-separated file, open with spreadsheet software">TAB file</a>

# min quality for phylogenetic evaluations = 50%

# ranking pairs of primers...

<img  src="http://floresta.eead.csic.es/primers4clades///jobs/26e8c759/26e8c759_intronless_cluster_0.fna_primers.list.oligo_eval.tab.png" alt="physical amplicon map" title="downloadable physical amplicon map">
<a href="http://floresta.eead.csic.es/primers4clades///jobs/26e8c759/26e8c759_intronless_cluster_0.fna_primers.list.oligo_eval.tab.png">download</a>

## Amplicon 1 codon_usage_table = input_derived_4e6be5b3 :
<b>GATAATTATGATTTAAAAGAAGAATTAggnaarggngc 5'->3' N 13 81</b> (aligned residues)
gataattacgatcttaaagaagaacttgggaagggagc >001 [Bombyx mori] Seq1 
gataactacgacctaaaagaagaattaggaaaaggagc >002 [Bombyx mori] Seq2 
gacaactatgaacttaaggaagaattaggcaagggcgc >003 [Bombyx mori] Seq3 
gacaattacgaccttaaggaagaattaggaaagggagc >004 [Bombyx mori] Seq4 
gataactatgaccttaaagaagaacttggaaaaggagc >005 [Bombyx mori] Seq5 
gataactacgatttgaaagaagaattaggcaaaggagc >006 [Bombyx mori] Seq6 
gataattatgatctgaaagaagaattaggaaaaggtgc >007 [Bombyx mori] Seq7 
gataattatgatcttaaagaagaattgggaaagggcgc >008 [Bombyx mori] Seq8 
..?..?..?..??.?..?......?.?..?.....?..
GATAATTATGATTTAAAAGAAGAATTAggvaargghgc codeh_corr 26e8c759_intronless_cluster_01_N13
GATAATTATGATTTAAAAGAAGAATTAggvaargghgc relax_corr 26e8c759_intronless_cluster_01_N13
gayaaytaygahytdaargaagaaytdggvaargghgc degen_corr 26e8c759_intronless_cluster_01_N13

<b>TCAAAAACTAAATAGTGAAAGTGTtcytcytgdat 5'->3' C 13 81</b> (aligned residues)
tcaaaaacgaggtagtgaaaatgttcctcttgaat >001 [Bombyx mori] Seq1 
tcaaatacgaggtaatgaaaatgttcttcttgaat >002 [Bombyx mori] Seq2 
tcaaagacaaggtaatggcaatgttcttcttgaat >003 [Bombyx mori] Seq3 
tcaaatacgagatagtggaagtgttcttcctgaat >004 [Bombyx mori] Seq4 
tcgaaaacgagataatggaaatgttcttcttgtat >005 [Bombyx mori] Seq5 
tcaaatacgaggtagtgaaaatgttcttcttgaat >006 [Bombyx mori] Seq6 
tcaaataccaaataatgaaaatgctcttcttgaat >007 [Bombyx mori] Seq7 
tcgaaaacaaggtaatgaaaatgttcctcttgaat >008 [Bombyx mori] Seq8 
..?..?..?.??..?..??.?..?........?..
TCAAAAACTAAATAGTGAAAGTGTtcytcytgwat codeh_corr 26e8c759_intronless_cluster_01_C81
TCAAAAACTAAATAGTGAAAGtgytcytcytgwat relax_corr 26e8c759_intronless_cluster_01_C81
tcraadacvarrtartgrmartgytcytcytgwat degen_corr 26e8c759_intronless_cluster_01_C81

# primer pair <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#quality" target="_new" title="estimation of primer quality based on 11 thermodynamic tests">quality</a> = 80%
# expected PCR product length (nt) = 207
# fwd: minTm = 52.2 maxTm = 56.5
# rev: minTm = 52.6 maxTm = 55.3
# quality <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#warnings" target="_new" title="tests failed by this pair of primers">warnings</a>:
# fwd fulldeg (31104  >= 193)
# rev fulldeg (18432 >= 193)
#--
# phylogenetic amplicon evaluation:
# subs.model = TIMIG
# alpha = 0.43
# mean <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#alRT" target="_new" title="approximate likelihood ratio test for branchs">aLRT</a> = 0.66 median aLRT = 0.76
# amplicon <a href="http://floresta.eead.csic.es/primers4clades/showjobfile.cgi?my_26e8c759_intronless_cluster_0_1__input_derived_4e6be5b3.dna_amp" target="_new" title="amplicon alignment in FASTA format">FASTA file</a>
# aLRT likelihood <a href="http://floresta.eead.csic.es/primers4clades///jobs/26e8c759/my_26e8c759_intronless_cluster_0_1__input_derived_4e6be5b3.png" target="_new" title="PNG figure for quick view of tree with aLRT support values">tree</a> (<a href="http://floresta.eead.csic.es/primers4clades///jobs/26e8c759/my_26e8c759_intronless_cluster_0_1__input_derived_4e6be5b3.phy_phyml_tree.txt" target="_new" title="tree with aLRT support values in Newick format for further analysis">newick</a>)
# end_of_amplicon



## Amplicon 2 codon_usage_table = input_derived_4e6be5b3 :
<b>AGAATCTGTTCACCACTGTcaycayaaygg 5'->3' N 122 188</b> (aligned residues)
agaatcagtacaccactgccaccataatgg >001 [Bombyx mori] Seq1 
agaatctgtccaccactgtcatcacaatgg >002 [Bombyx mori] Seq2 
agaatctgtccatcactgtcaccataatgg >003 [Bombyx mori] Seq3 
ggaatctgttcatcattgccatcataatgg >004 [Bombyx mori] Seq4 
ggaatccgtgcaccattgtcaccacaatgg >005 [Bombyx mori] Seq5 
ggaatccgtacatcattgtcaccacaatgg >006 [Bombyx mori] Seq6 
ggagtcagtgcatcactgccaccataatgg >007 [Bombyx mori] Seq7 
tgaatctgtacaccattgtcaccataatgg >008 [Bombyx mori] Seq8 
?..?..?..?..?..?..?........!..
AGAATCTGTTCACCACTGTcaycayaatgg codeh_corr 26e8c759_intronless_cluster_02_N122
AGAATCTGTTCACcaytgycaycayaatgg relax_corr 26e8c759_intronless_cluster_02_N122
dgartchgtncaycaytgycaycayaatgg degen_corr 26e8c759_intronless_cluster_02_N122

<b>AATATCAACTGGTTTACCATATggytcyttytt 5'->3' C 122 188</b> (aligned residues)
aatatctactggcttaccgtaaggctctttttt >001 [Bombyx mori] Seq1 
aatgtctacgggctttccgtacggctctttttt >002 [Bombyx mori] Seq2 
tatatccactggtttaccataaggttccttttt >003 [Bombyx mori] Seq3 
aatatctacgggctttccataaggctctttctt >004 [Bombyx mori] Seq4 
aatatccactggcttgccgtaaggttctttttt >005 [Bombyx mori] Seq5 
tatatcaactggtttgccataaggttctttttt >006 [Bombyx mori] Seq6 
aatatctactggtttgccgtatggctccttctt >007 [Bombyx mori] Seq7 
gatatctactggtttaccgtatggttctttttt >008 [Bombyx mori] Seq8 
?..?..?..?..?..?..?..?...........
AATATCAACTGGTTTACCATATggytcyttytt codeh_corr 26e8c759_intronless_cluster_02_C188
AATATCAACTGGTTTACCAtahggytcyttytt relax_corr 26e8c759_intronless_cluster_02_C188
datrtchackggyttdccrtahggytcyttytt degen_corr 26e8c759_intronless_cluster_02_C188

# primer pair <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#quality" target="_new" title="estimation of primer quality based on 11 thermodynamic tests">quality</a> = 80%
# expected PCR product length (nt) = 201
# fwd: minTm = 58.1 maxTm = 61.7
# rev: minTm = 54.5 maxTm = 57.8
# quality <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#warnings" target="_new" title="tests failed by this pair of primers">warnings</a>:
# fwd fulldeg (2304  >= 193)
# rev fulldeg (10368 >= 193)
#--
# phylogenetic amplicon evaluation:
# subs.model = TrNIG
# alpha = 0.56
# mean <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#alRT" target="_new" title="approximate likelihood ratio test for branchs">aLRT</a> = 0.52 median aLRT = 0.79
# amplicon <a href="http://floresta.eead.csic.es/primers4clades/showjobfile.cgi?my_26e8c759_intronless_cluster_0_2__input_derived_4e6be5b3.dna_amp" target="_new" title="amplicon alignment in FASTA format">FASTA file</a>
# aLRT likelihood <a href="http://floresta.eead.csic.es/primers4clades///jobs/26e8c759/my_26e8c759_intronless_cluster_0_2__input_derived_4e6be5b3.png" target="_new" title="PNG figure for quick view of tree with aLRT support values">tree</a> (<a href="http://floresta.eead.csic.es/primers4clades///jobs/26e8c759/my_26e8c759_intronless_cluster_0_2__input_derived_4e6be5b3.phy_phyml_tree.txt" target="_new" title="tree with aLRT support values in Newick format for further analysis">newick</a>)
# end_of_amplicon



## Amplicon 3 codon_usage_table = input_derived_4e6be5b3 :
<b>TGAAGTTCAAGGTGATCAAcargcntggtt 5'->3' N 163 283</b> (aligned residues)
tgaagtgcaaggagaacagcaggcctggtt >001 [Bombyx mori] Seq1 
cgaggtgcaaggagaccagcaagcatggtt >002 [Bombyx mori] Seq2 
cgaagttcaaggagatcaacaagcatggtt >003 [Bombyx mori] Seq3 
tgaagtccaaggtgaacaacaagcgtggtt >004 [Bombyx mori] Seq4 
cgaagttcaaggggaccagcaagcgtggtt >005 [Bombyx mori] Seq5 
tgaagtccaaggtgatcaacaagcatggtt >006 [Bombyx mori] Seq6 
agaagtgcaaggtgaccaacaggcgtggtt >007 [Bombyx mori] Seq7 
tgaagtacaaggcgaccagcaagcttggtt >008 [Bombyx mori] Seq8 
?..?..?.....?..?..?...........
TGAAGTTCAAGGTGATCAAcargcntggtt codeh_corr 26e8c759_intronless_cluster_03_N163
TGAAGTTCAAGGTGATcarcargcntggtt relax_corr 26e8c759_intronless_cluster_03_N163
hgargtncaaggngahcarcargcntggtt degen_corr 26e8c759_intronless_cluster_03_N163

<b>CTTTAAACAATCAACAGTTTCTtgnckrtgcat 5'->3' C 163 283</b> (aligned residues)
cttcaagcagtccacagtctcctgcctgtgcat >001 [Bombyx mori] Seq1 
cttcaagcaatccactgtttcttgcctgtgcat >002 [Bombyx mori] Seq2 
tttcaaacaatccacagtttcctgtctatgcat >003 [Bombyx mori] Seq3 
cttcaagcaatccacagtctcctgtctgtgcat >004 [Bombyx mori] Seq4 
tttcaaacagtctacagtctcctggcggtgcat >005 [Bombyx mori] Seq5 
cttcaagcagtccactgtttcctgtctatgcat >006 [Bombyx mori] Seq6 
cttcaagcagtctactgtctcttgtctgtgcat >007 [Bombyx mori] Seq7 
cttcaaacagtccacggtctcttgcctgtgcat >008 [Bombyx mori] Seq8 
?..!..?..?..?..?..?..?..?........
CTTCAAACAATCAACAGTTTCTtgbckrtgcat codeh_corr 26e8c759_intronless_cluster_03_C283
CTTCAAACAATCAACAGTTtcytgbckrtgcat relax_corr 26e8c759_intronless_cluster_03_C283
yttcaarcartcyacdgtytcytgbckrtgcat degen_corr 26e8c759_intronless_cluster_03_C283

# primer pair <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#quality" target="_new" title="estimation of primer quality based on 11 thermodynamic tests">quality</a> = 80%
# expected PCR product length (nt) = 363
# fwd: minTm = 60.1 maxTm = 63.0
# rev: minTm = 56.8 maxTm = 61.9
# quality <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#warnings" target="_new" title="tests failed by this pair of primers">warnings</a>:
# fwd fulldeg (4608  >= 193)
# rev fulldeg (2304 >= 193)
#--
# phylogenetic amplicon evaluation:
# subs.model = TIMIG
# alpha = 0.25
# mean <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#alRT" target="_new" title="approximate likelihood ratio test for branchs">aLRT</a> = 0.46 median aLRT = 0.42
# amplicon <a href="http://floresta.eead.csic.es/primers4clades/showjobfile.cgi?my_26e8c759_intronless_cluster_0_3__input_derived_4e6be5b3.dna_amp" target="_new" title="amplicon alignment in FASTA format">FASTA file</a>
# aLRT likelihood <a href="http://floresta.eead.csic.es/primers4clades///jobs/26e8c759/my_26e8c759_intronless_cluster_0_3__input_derived_4e6be5b3.png" target="_new" title="PNG figure for quick view of tree with aLRT support values">tree</a> (<a href="http://floresta.eead.csic.es/primers4clades///jobs/26e8c759/my_26e8c759_intronless_cluster_0_3__input_derived_4e6be5b3.phy_phyml_tree.txt" target="_new" title="tree with aLRT support values in Newick format for further analysis">newick</a>)
# end_of_amplicon



## Amplicon 4 codon_usage_table = Bombyx_mori :
<b>GACCTGAAAGAAGAACTGggnaarggngc 5'->3' N 16 172</b> (aligned residues)
gatcttaaagaagaacttgggaagggagc >001 [Bombyx mori] Seq1 
gacctaaaagaagaattaggaaaaggagc >002 [Bombyx mori] Seq2 
gaacttaaggaagaattaggcaagggcgc >003 [Bombyx mori] Seq3 
gaccttaaggaagaattaggaaagggagc >004 [Bombyx mori] Seq4 
gaccttaaagaagaacttggaaaaggagc >005 [Bombyx mori] Seq5 
gatttgaaagaagaattaggcaaaggagc >006 [Bombyx mori] Seq6 
gatctgaaagaagaattaggaaaaggtgc >007 [Bombyx mori] Seq7 
gatcttaaagaagaattgggaaagggcgc >008 [Bombyx mori] Seq8 
..??.?..?......?.?..?.....?..
GACCTGAAAGAAGAACTGggvaargghgc codeh_corr 26e8c759_intronless_cluster_01_N16
GACCTGAAAGAAGAACTGggvaargghgc relax_corr 26e8c759_intronless_cluster_01_N16
gahytdaargaagaaytdggvaargghgc degen_corr 26e8c759_intronless_cluster_01_N16

<b>CCAGGTGTACCAGCGaanccraacca 5'->3' C 16 172</b> (aligned residues)
cctggtgttccagcaaatccaaacca >001 [Bombyx mori] Seq1 
cccggagtgcctgcgaatccaaacca >002 [Bombyx mori] Seq2 
ccgggtgtacctgcgaatccgaacca >003 [Bombyx mori] Seq3 
cctggagtaccagcaaagccaaacca >004 [Bombyx mori] Seq4 
ccaggagtaccagcaaatccgaacca >005 [Bombyx mori] Seq5 
ccaggggtaccggcaaatccaaacca >006 [Bombyx mori] Seq6 
ccaggagtaccagcgaaaccgaacca >007 [Bombyx mori] Seq7 
ccaggagtacctgcaaaaccaaacca >008 [Bombyx mori] Seq8 
..?..?..?..?..?..?........
CCAGGTGTACCAGCGaadccraacca codeh_corr 26e8c759_intronless_cluster_01_C172
CCAGGTGTACCAgcraadccraacca relax_corr 26e8c759_intronless_cluster_01_C172
ccnggdgtdccdgcraadccraacca degen_corr 26e8c759_intronless_cluster_01_C172

# primer pair <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#quality" target="_new" title="estimation of primer quality based on 11 thermodynamic tests">quality</a> = 80%
# expected PCR product length (nt) = 471
# fwd: minTm = 62.0 maxTm = 67.5
# rev: minTm = 65.5 maxTm = 68.8
# quality <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#warnings" target="_new" title="tests failed by this pair of primers">warnings</a>:
# fwd fulldeg (3888  >= 193)
# rev fulldeg (1296 >= 193)
#--
# phylogenetic amplicon evaluation:
# subs.model = TIMIG
# alpha = 0.24
# mean <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#alRT" target="_new" title="approximate likelihood ratio test for branchs">aLRT</a> = 0.49 median aLRT = 0.57
# amplicon <a href="http://floresta.eead.csic.es/primers4clades/showjobfile.cgi?my_26e8c759_intronless_cluster_0_1__Bombyx_mori.dna_amp" target="_new" title="amplicon alignment in FASTA format">FASTA file</a>
# aLRT likelihood <a href="http://floresta.eead.csic.es/primers4clades///jobs/26e8c759/my_26e8c759_intronless_cluster_0_1__Bombyx_mori.png" target="_new" title="PNG figure for quick view of tree with aLRT support values">tree</a> (<a href="http://floresta.eead.csic.es/primers4clades///jobs/26e8c759/my_26e8c759_intronless_cluster_0_1__Bombyx_mori.phy_phyml_tree.txt" target="_new" title="tree with aLRT support values in Newick format for further analysis">newick</a>)
# end_of_amplicon



## Amplicon 5 codon_usage_table = Bombyx_mori :
<b>ACGACTACCCTTCACCTgartgggayac 5'->3' N 232 320</b> (aligned residues)
atgattatccttctcctgagtgggacac >001 [Bombyx mori] Seq1 
atgactacccctccccagagtgggacac >002 [Bombyx mori] Seq2 
atgattatccgtctcctgagtgggatac >003 [Bombyx mori] Seq3 
atgattacccatcacctgaatgggatac >004 [Bombyx mori] Seq4 
atgattacccgtcaccggaatgggacac >005 [Bombyx mori] Seq5 
acgattacccatctcccgaatgggatac >006 [Bombyx mori] Seq6 
acgattatccctctcctgaatgggatac >007 [Bombyx mori] Seq7 
atgattatccatcacccgagtgggatac >008 [Bombyx mori] Seq8 
.?..?..?..?..?..?...........
ACGACTACCCTTCACCTgartgggayac codeh_corr 26e8c759_intronless_cluster_02_N232
ACGACTACCCTTCAccngartgggayac relax_corr 26e8c759_intronless_cluster_02_N232
aygaytayccntchccngartgggayac degen_corr 26e8c759_intronless_cluster_02_N232

<b>TGAACCGTCACCCTTTttrttnacnat 5'->3' C 232 320</b> (aligned residues)
agatccatcactctttttattaaccat >001 [Bombyx mori] Seq1 
agaaccgtccccttttttgttaactat >002 [Bombyx mori] Seq2 
cgatccatcgccttttttattaaccat >003 [Bombyx mori] Seq3 
tgatccttcacttttcttgttaaccat >004 [Bombyx mori] Seq4 
agaaccatcacctttcttattaaccat >005 [Bombyx mori] Seq5 
tgaaccatcacctttcttattgaccat >006 [Bombyx mori] Seq6 
cgatccatcaccttttttgttgaccat >007 [Bombyx mori] Seq7 
tgacccatctccttttttgttcaccat >008 [Bombyx mori] Seq8 
?..?..?..?.??..?.....?..?..
TGAACCGTCACCCTTTttrttvacyat codeh_corr 26e8c759_intronless_cluster_02_C320
TGAACCGTCACCCttyttrttvacyat relax_corr 26e8c759_intronless_cluster_02_C320
hgahccdtcncyyttyttrttvacyat degen_corr 26e8c759_intronless_cluster_02_C320

# primer pair <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#quality" target="_new" title="estimation of primer quality based on 11 thermodynamic tests">quality</a> = 80%
# expected PCR product length (nt) = 267
# fwd: minTm = 62.1 maxTm = 65.5
# rev: minTm = 56.1 maxTm = 62.5
# quality <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#warnings" target="_new" title="tests failed by this pair of primers">warnings</a>:
# fwd fulldeg (1536  >= 193)
# rev fulldeg (10368 >= 193)
#--
# phylogenetic amplicon evaluation:
# subs.model = TVMI
# alpha = 0
# mean <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#alRT" target="_new" title="approximate likelihood ratio test for branchs">aLRT</a> = 0.75 median aLRT = 0.81
# amplicon <a href="http://floresta.eead.csic.es/primers4clades/showjobfile.cgi?my_26e8c759_intronless_cluster_0_2__Bombyx_mori.dna_amp" target="_new" title="amplicon alignment in FASTA format">FASTA file</a>
# aLRT likelihood <a href="http://floresta.eead.csic.es/primers4clades///jobs/26e8c759/my_26e8c759_intronless_cluster_0_2__Bombyx_mori.png" target="_new" title="PNG figure for quick view of tree with aLRT support values">tree</a> (<a href="http://floresta.eead.csic.es/primers4clades///jobs/26e8c759/my_26e8c759_intronless_cluster_0_2__Bombyx_mori.phy_phyml_tree.txt" target="_new" title="tree with aLRT support values in Newick format for further analysis">newick</a>)
# end_of_amplicon



## Amplicon 6 codon_usage_table = Bombyx_mori :
<b>AAACAGTGGACTGCCTGAAraarttyaayg 5'->3' N 287 370</b> (aligned residues)
agactgtggactgcttgaagaagttcaatg >001 [Bombyx mori] Seq1 
aaacagtggattgcttgaagaaattcaatg >002 [Bombyx mori] Seq2 
aaactgtggattgtttgaaaaaattcaatg >003 [Bombyx mori] Seq3 
agactgtggattgcttgaagaaattcaatg >004 [Bombyx mori] Seq4 
agactgtagactgtttgaaaaagttcaatg >005 [Bombyx mori] Seq5 
aaacagtggactgcttgaagaaattcaacg >006 [Bombyx mori] Seq6 
agacagtagactgcttgaagaaattcaatg >007 [Bombyx mori] Seq7 
agaccgtggactgtttgaagaaattcaatg >008 [Bombyx mori] Seq8 
.?..?..?..?..?!..........!....
AAACAGTGGACTGCTTGAAraarttcaayg codeh_corr 26e8c759_intronless_cluster_03_N287
AAACAGTGGACtgyttgaaraarttcaayg relax_corr 26e8c759_intronless_cluster_03_N287
arachgtrgaytgyttgaaraarttcaayg degen_corr 26e8c759_intronless_cluster_03_N287

<b>TCACCTCTAGCTTTACCGAATswrtcngcytt 5'->3' C 287 370</b> (aligned residues)
tctcctcttgtttttccaaacgaatctgcttt >001 [Bombyx mori] Seq1 
tcaccgcgagctttgccgaaggaatcagcttt >002 [Bombyx mori] Seq2 
tcgccccgagcttttccaaacgagtccgcttt >003 [Bombyx mori] Seq3 
tcgcctcgagcttttccaaaagaatcagcttt >004 [Bombyx mori] Seq4 
tcgccgcgagcttttccaaacgaatctgcctt >005 [Bombyx mori] Seq5 
tcacctcgagcctttccgaatgagtctgcttt >006 [Bombyx mori] Seq6 
tcgccacgcgcttttccaaaagagtctgcttt >007 [Bombyx mori] Seq7 
tcaccacgagctttaccgaatgaatcagcttt >008 [Bombyx mori] Seq8 
..?..?.??.??..?..?..?!!...?.....
TCACCTCTAGCTTTACCGAATgartchgcytt codeh_corr 26e8c759_intronless_cluster_03_C370
TCACCTCTAGCTTTACCGAATgartchgcytt relax_corr 26e8c759_intronless_cluster_03_C370
tcdccnckhgyyttdccraangartchgcytt degen_corr 26e8c759_intronless_cluster_03_C370

# primer pair <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#quality" target="_new" title="estimation of primer quality based on 11 thermodynamic tests">quality</a> = 80%
# expected PCR product length (nt) = 252
# fwd: minTm = 56.9 maxTm = 60.5
# rev: minTm = 59.5 maxTm = 63.4
# quality <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#warnings" target="_new" title="tests failed by this pair of primers">warnings</a>:
# fwd fulldeg (384  >= 193)
# rev fulldeg (82944 >= 193)
#--
# phylogenetic evaluation skipped (max runtime 300s)
# end_of_amplicon



## Amplicon 7 codon_usage_table = Bombyx_mori :
<b>CGCTATCAACAACGGTgaytayganac 5'->3' N 399 467</b> (aligned residues)
tgctattaataacggcgattatgaaac >001 [Bombyx mori] Seq1 
tgctataaataacggcgattatgatac >002 [Bombyx mori] Seq2 
tgcaatcaataatggtgattacgacac >003 [Bombyx mori] Seq3 
cgctattaacaatggcgactatgagac >004 [Bombyx mori] Seq4 
cgcaatcaataatggcgactacgacac >005 [Bombyx mori] Seq5 
tgctattaataacggagactatgaaac >006 [Bombyx mori] Seq6 
cgctattaacaacggtgattacgatac >007 [Bombyx mori] Seq7 
tgcaattaataacggtgattatgaaac >008 [Bombyx mori] Seq8 
?..?..?..?..?..?...........
CGCTATCAACAACGGTgaytayganac codeh_corr 26e8c759_intronless_cluster_04_N399
CGCTATCAACAACGGTgaytayganac relax_corr 26e8c759_intronless_cluster_04_N399
ygcwathaayaaygghgaytayganac degen_corr 26e8c759_intronless_cluster_04_N399

<b>TGATTGTGTCACGCACacrtangcdat 5'->3' C 399 467</b> (aligned residues)
actttgtgtgacgcatacgtacgctat >001 [Bombyx mori] Seq1 
actttgtgttacgcagacatatgctat >002 [Bombyx mori] Seq2 
gctttgtgttacgcagacgtatgctat >003 [Bombyx mori] Seq3 
actttgtgtaacacacacatatgcaat >004 [Bombyx mori] Seq4 
gctctgtgtaacgcacacataagctat >005 [Bombyx mori] Seq5 
actctgcgttacacacacgtaagctat >006 [Bombyx mori] Seq6 
actctgtgtcacgcatacatatgctat >007 [Bombyx mori] Seq7 
actttgtgttacgcatacgtatgcgat >008 [Bombyx mori] Seq8 
?!!?..?..?..?..?.....?.....
TCTTTGTGTCACGCACacrtahgcdat codeh_corr 26e8c759_intronless_cluster_04_C467
TCTTTGTGTCACGCACacrtahgcdat relax_corr 26e8c759_intronless_cluster_04_C467
rctytgygtnacrcabacrtahgcdat degen_corr 26e8c759_intronless_cluster_04_C467

# primer pair <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#quality" target="_new" title="estimation of primer quality based on 11 thermodynamic tests">quality</a> = 80%
# expected PCR product length (nt) = 207
# fwd: minTm = 57.6 maxTm = 62.5
# rev: minTm = 59.4 maxTm = 64.4
# quality <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#warnings" target="_new" title="tests failed by this pair of primers">warnings</a>:
# fwd fulldeg (2304  >= 193)
# rev fulldeg (3456 >= 193)
#--
# phylogenetic evaluation skipped (max runtime 300s)
# end_of_amplicon



## Amplicon 8 codon_usage_table = input_derived_4e6be5b3 :
<b>CAAGAAACTGTTGATTGTTTGAAraarttyaayg 5'->3' N 286 433</b> (aligned residues)
caggagactgtggactgcttgaagaagttcaatg >001 [Bombyx mori] Seq1 
caagaaacagtggattgcttgaagaaattcaatg >002 [Bombyx mori] Seq2 
caggaaactgtggattgtttgaaaaaattcaatg >003 [Bombyx mori] Seq3 
caggagactgtggattgcttgaagaaattcaatg >004 [Bombyx mori] Seq4 
caggagactgtagactgtttgaaaaagttcaatg >005 [Bombyx mori] Seq5 
caggaaacagtggactgcttgaagaaattcaacg >006 [Bombyx mori] Seq6 
caagagacagtagactgcttgaagaaattcaatg >007 [Bombyx mori] Seq7 
caagagaccgtggactgtttgaagaaattcaatg >008 [Bombyx mori] Seq8 
..?..?..?..?..?..?...........!....
CAAGAAACTGTTGATTGTTTGAAraarttcaayg codeh_corr 26e8c759_intronless_cluster_04_N286
CAAGAAACTGTTGATtgyttgaaraarttcaayg relax_corr 26e8c759_intronless_cluster_04_N286
cargarachgtrgaytgyttgaaraarttcaayg degen_corr 26e8c759_intronless_cluster_04_N286

<b>TTGGAGTATTATCAATGAAAAATttrtgraaytc 5'->3' C 286 433</b> (aligned residues)
tagccgtattatctataaagaatttatggaattc >001 [Bombyx mori] Seq1 
tcggtgtgttgtcgataaagaatttatgaaattc >002 [Bombyx mori] Seq2 
gcggtgtattgtcaataaagaatttatggaattc >003 [Bombyx mori] Seq3 
ttggtgtattatcgatgaaaaacttgtgaaattc >004 [Bombyx mori] Seq4 
tcggagtgttatcaataaagaacttgtgaaattc >005 [Bombyx mori] Seq5 
taggagtattatcgatgaagaatttatgaaactc >006 [Bombyx mori] Seq6 
ccggagtgttgtcaatgaagaatttatggaattc >007 [Bombyx mori] Seq7 
ttggcgtgttatcgataaagaacttgtggaactc >008 [Bombyx mori] Seq8 
??.??..?..?..?..?..?..?...........
TTGGAGTATTATCAATGAAAAATttrtgraaytc codeh_corr 26e8c759_intronless_cluster_04_C433
TTGGAGTATTATCAATGAAAaayttrtgraaytc relax_corr 26e8c759_intronless_cluster_04_C433
bhgshgtrttrtcdatraaraayttrtgraaytc degen_corr 26e8c759_intronless_cluster_04_C433

# primer pair <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#quality" target="_new" title="estimation of primer quality based on 11 thermodynamic tests">quality</a> = 60%
# expected PCR product length (nt) = 444
# fwd: minTm = 54.1 maxTm = 57.3
# rev: minTm = 50.3 maxTm = 54.3
# quality <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#warnings" target="_new" title="tests failed by this pair of primers">warnings</a>:
# fwd fulldeg (768  >= 193)
# fwd hpinpot (0.62 >= 0.61)
# rev fulldeg (41472 >= 193)
# rev hpinpot (0.72 >= 0.61)
#--
# phylogenetic evaluation skipped (max runtime 300s)
# end_of_amplicon






## abbreviations:
# quality     = quality estimation [best = 100, worst = 0] %
# crosspot    = potential of cross-hybridization [0-1]
# relaxdeg    = relaxed (3' extended segment) degeneracy
# fulldeg     = full primer degeneracy 
# minTm       = minimum Tm for the pool of relaxed primers
# maxTm       = maximum Tm for the pool of relaxed primers
# hpinpot     = potential of primer hairpin [0-1]
# selfpot     = potential of primer self-priming [0-1]
# aLRT        = approximate likelihood-ratio test [0-1],
#               overall confidence on trees built using this amplicon
