<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>EvolTree class &mdash; A Python Environment for (phylogenetic) Tree Exploration</title>
    
    <link rel="stylesheet" href="../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '2.3.10',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="top" title="A Python Environment for (phylogenetic) Tree Exploration" href="../index.html" />
    <link rel="up" title="ETE’s Reference Guide" href="index.html" />
    <link rel="next" title="NCBITaxa class" href="reference_ncbi.html" />
    <link rel="prev" title="WebTreeApplication object" href="reference_webplugin.html" />
<script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
           (i[r].q=i[r].q||[]).push(arguments)
    },i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

 ga('create', 'UA-783841-4', 'etetoolkit.org');
 ga('send', 'pageview');
 </script>

<!-- 
 <script type="text/javascript">
 var _gaq = _gaq || [];
 _gaq.push(['_setAccount', 'UA-783841-12']);
 _gaq.push(['_setCookiePath', '/ete2/']);
 _gaq.push(['_trackPageview']);

 (function() {
 var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
 ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
 var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
 })();
 </script>
 -->
<!-- <link rel="canonical" href="http://etetoolkit.org/docs/2.3/CANONICALPATH" />
<meta http-equiv="refresh" content="0; url=http:/etetoolkit.org/CANONICALPATH" />
<script type="text/javascript">
window.location.replace('http://etetoolkit.org/docs/2.3/' + window.location.pathname.replace('/ete2/',''));
</script> -->
 
 

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="reference_ncbi.html" title="NCBITaxa class"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="reference_webplugin.html" title="WebTreeApplication object"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">ETE&#8217;s Reference Guide</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-ete2.evol"></span><div class="section" id="evoltree-class">
<h1>EvolTree class<a class="headerlink" href="#evoltree-class" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="ete2.EvolNode">
<em class="property">class </em><code class="descname">EvolNode</code><span class="sig-paren">(</span><em>newick=None</em>, <em>alignment=None</em>, <em>alg_format='fasta'</em>, <em>sp_naming_function=&lt;function _parse_species&gt;</em>, <em>format=0</em>, <em>binpath=''</em><span class="sig-paren">)</span><a class="headerlink" href="#ete2.EvolNode" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">ete2.phylo.phylotree.PhyloNode</span></code></p>
<p>Re-implementation of the standart TreeNode instance. It adds
attributes and methods to work with phylogentic trees.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>newick</strong> &#8211; path to tree in newick format, can also be a string</li>
<li><strong>alignment</strong> &#8211; path to alignment, can also be a string.</li>
<li><strong>alg_format</strong> (<em>fasta</em>) &#8211; alignment format.</li>
<li><strong>sp_naming_function</strong> &#8211; function to infer species name.</li>
<li><strong>format</strong> &#8211; type of newick format</li>
<li><strong>binpath</strong> &#8211; path to binaries, in case codeml or SLR are not in global path.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="ete2.EvolNode.change_dist_to_evol">
<code class="descname">change_dist_to_evol</code><span class="sig-paren">(</span><em>evol</em>, <em>model</em>, <em>fill=False</em><span class="sig-paren">)</span><a class="headerlink" href="#ete2.EvolNode.change_dist_to_evol" title="Permalink to this definition">¶</a></dt>
<dd><p>change dist/branch length of the tree to a given evolutionary
variable (dN, dS, w or bL), default is bL.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>evol</strong> &#8211; evolutionary variable</li>
<li><strong>model</strong> &#8211; Model object from which to retrieve evolutionary variables</li>
<li><strong>fill</strong> (<em>False</em>) &#8211; do not affects only dist parameter, each node will be annotated with all evolutionary variables (nodel.dN, node.w...).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="ete2.EvolNode.get_descendant_by_node_id">
<code class="descname">get_descendant_by_node_id</code><span class="sig-paren">(</span><em>idname</em><span class="sig-paren">)</span><a class="headerlink" href="#ete2.EvolNode.get_descendant_by_node_id" title="Permalink to this definition">¶</a></dt>
<dd><p>returns node list corresponding to a given idname</p>
</dd></dl>

<dl class="method">
<dt id="ete2.EvolNode.get_evol_model">
<code class="descname">get_evol_model</code><span class="sig-paren">(</span><em>modelname</em><span class="sig-paren">)</span><a class="headerlink" href="#ete2.EvolNode.get_evol_model" title="Permalink to this definition">¶</a></dt>
<dd><p>returns one precomputed model</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>modelname</strong> &#8211; string of the name of a model object stored</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Model object</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="ete2.EvolNode.get_most_likely">
<code class="descname">get_most_likely</code><span class="sig-paren">(</span><em>altn</em>, <em>null</em><span class="sig-paren">)</span><a class="headerlink" href="#ete2.EvolNode.get_most_likely" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns pvalue of LRT between alternative model and null model.</p>
<p>usual comparison are:</p>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="11%" />
<col width="69%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Alternative</th>
<th class="head">Null</th>
<th class="head">Test</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>M2</td>
<td>M1</td>
<td>PS on sites (M2 prone to miss some sites)
(Yang 2000)</td>
</tr>
<tr class="row-odd"><td>M3</td>
<td>M0</td>
<td>test of variability among sites</td>
</tr>
<tr class="row-even"><td>M8</td>
<td>M7</td>
<td>PS on sites
(Yang 2000)</td>
</tr>
<tr class="row-odd"><td>M8</td>
<td>M8a</td>
<td>RX on sites?? think so....</td>
</tr>
<tr class="row-even"><td>bsA</td>
<td>bsA1</td>
<td>PS on sites on specific branch
(Zhang 2005)</td>
</tr>
<tr class="row-odd"><td>bsA</td>
<td>M1</td>
<td>RX on sites on specific branch
(Zhang 2005)</td>
</tr>
<tr class="row-even"><td>bsC</td>
<td>M1</td>
<td>different omegas on clades branches sites
ref: Yang Nielsen 2002</td>
</tr>
<tr class="row-odd"><td>bsD</td>
<td>M3</td>
<td>different omegas on clades branches sites
(Yang Nielsen 2002, Bielawski 2004)</td>
</tr>
<tr class="row-even"><td>b_free</td>
<td>b_neut</td>
<td><dl class="first last docutils">
<dt>foreground branch not neutral (w != 1)</dt>
<dd><ul class="first simple">
<li>RX if P&lt;0.05 (means that w on frg=1)</li>
<li>PS if P&gt;0.05 and wfrg&gt;1</li>
<li>CN if P&gt;0.05 and wfrg&gt;1</li>
</ul>
<p class="last">(Yang Nielsen 2002)</p>
</dd>
</dl>
</td>
</tr>
<tr class="row-odd"><td>b_free</td>
<td>M0</td>
<td>different ratio on branches
(Yang Nielsen 2002)</td>
</tr>
</tbody>
</table>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>altn</strong> &#8211; model with higher number of parameters (np)</li>
<li><strong>null</strong> &#8211; model with lower number of parameters (np)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="ete2.EvolNode.link_to_alignment">
<code class="descname">link_to_alignment</code><span class="sig-paren">(</span><em>alignment</em>, <em>alg_format='paml'</em>, <em>nucleotides=True</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ete2.EvolNode.link_to_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>same function as for phyloTree, but translate sequences if nucleotides
nucleotidic sequence is kept under node.nt_sequence</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>alignment</strong> (<em>True</em>) &#8211; path to alignment or string</li>
<li><strong>alg_format</strong> &#8211; one of fasta phylip or paml</li>
<li><strong>alignment</strong> &#8211; set to False in case we want to keep it untranslated</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="ete2.EvolNode.link_to_evol_model">
<code class="descname">link_to_evol_model</code><span class="sig-paren">(</span><em>path</em>, <em>model</em><span class="sig-paren">)</span><a class="headerlink" href="#ete2.EvolNode.link_to_evol_model" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>link EvolTree to evolutionary model</dt>
<dd><ul class="first last simple">
<li>free-branch model (&#8216;fb&#8217;) will append evol values to tree</li>
<li>Site models (M0, M1, M2, M7, M8) will give evol values by site
and likelihood</li>
</ul>
</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>path</strong> &#8211; path to outfile containing model computation result</li>
<li><strong>model</strong> &#8211; either the name of a model, or a Model object (usually empty)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="ete2.EvolNode.mark_tree">
<code class="descname">mark_tree</code><span class="sig-paren">(</span><em>node_ids</em>, <em>verbose=False</em>, <em>**kargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ete2.EvolNode.mark_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>function to mark branches on tree in order that paml could interpret it.
takes a &#8220;marks&#8221; argument that should be a list of #1,#1,#2
e.g.:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">t</span><span class="o">=</span><span class="n">Tree</span><span class="o">.</span><span class="n">mark_tree</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="n">marks</span><span class="o">=</span><span class="p">[</span><span class="s">&quot;#1&quot;</span><span class="p">,</span><span class="s">&quot;#2&quot;</span><span class="p">])</span>
</pre></div>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>node_ids</strong> &#8211; list of node ids (have a look to node.node_id)</li>
<li><strong>verbose</strong> (<em>False</em>) &#8211; warn if marks do not correspond to codeml standard</li>
<li><strong>kargs</strong> &#8211; mainly for the marks key-word which needs a list of marks (marks=[&#8216;#1&#8217;, &#8216;#2&#8217;])</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="ete2.EvolNode.render">
<code class="descname">render</code><span class="sig-paren">(</span><em>file_name</em>, <em>layout=None</em>, <em>w=None</em>, <em>h=None</em>, <em>tree_style=None</em>, <em>header=None</em>, <em>histfaces=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ete2.EvolNode.render" title="Permalink to this definition">¶</a></dt>
<dd><p>call super show adding up and down faces</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>layout</strong> &#8211; a layout function</li>
<li><strong>tree_style</strong> (<em>None</em>) &#8211; tree_style object</li>
<li><strong>Nonehistface</strong> &#8211; an histogram face function. This is only to plot selective pressure among sites</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="ete2.EvolNode.run_model">
<code class="descname">run_model</code><span class="sig-paren">(</span><em>model_name</em>, <em>ctrl_string=''</em>, <em>keep=True</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ete2.EvolNode.run_model" title="Permalink to this definition">¶</a></dt>
<dd><p>To compute evolutionnary models.     e.g.: b_free_lala.vs.lele, will launch one free branch model, and store 
it in &#8220;WORK_DIR/b_free_lala.vs.lele&#8221; directory</p>
<p>WARNING: this functionality needs to create a working directory in &#8220;rep&#8221;</p>
<p>WARNING: you need to have codeml and/or SLR in your path</p>
<p>The models available are:</p>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="50%" />
<col width="31%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Model name</th>
<th class="head">Description</th>
<th class="head">Model kind</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>M1</td>
<td>relaxation</td>
<td>site</td>
</tr>
<tr class="row-odd"><td>M10</td>
<td>beta and gamma + 1</td>
<td>site</td>
</tr>
<tr class="row-even"><td>M11</td>
<td>beta and normal &gt; 1</td>
<td>site</td>
</tr>
<tr class="row-odd"><td>M12</td>
<td>0 and 2 normal &gt; 2</td>
<td>site</td>
</tr>
<tr class="row-even"><td>M13</td>
<td>3 normal &gt; 0</td>
<td>site</td>
</tr>
<tr class="row-odd"><td>M2</td>
<td>positive-selection</td>
<td>site</td>
</tr>
<tr class="row-even"><td>M3</td>
<td>discrete</td>
<td>site</td>
</tr>
<tr class="row-odd"><td>M4</td>
<td>frequencies</td>
<td>site</td>
</tr>
<tr class="row-even"><td>M5</td>
<td>gamma</td>
<td>site</td>
</tr>
<tr class="row-odd"><td>M6</td>
<td>2 gamma</td>
<td>site</td>
</tr>
<tr class="row-even"><td>M7</td>
<td>relaxation</td>
<td>site</td>
</tr>
<tr class="row-odd"><td>M8</td>
<td>positive-selection</td>
<td>site</td>
</tr>
<tr class="row-even"><td>M8a</td>
<td>relaxation</td>
<td>site</td>
</tr>
<tr class="row-odd"><td>M9</td>
<td>beta and gamma</td>
<td>site</td>
</tr>
<tr class="row-even"><td>SLR</td>
<td>positive/negative selection</td>
<td>site</td>
</tr>
<tr class="row-odd"><td>M0</td>
<td>negative-selection</td>
<td>null</td>
</tr>
<tr class="row-even"><td>fb_anc</td>
<td>free-ratios</td>
<td>branch_ancestor</td>
</tr>
<tr class="row-odd"><td>bsA</td>
<td>positive-selection</td>
<td>branch-site</td>
</tr>
<tr class="row-even"><td>bsA1</td>
<td>relaxation</td>
<td>branch-site</td>
</tr>
<tr class="row-odd"><td>bsB</td>
<td>positive-selection</td>
<td>branch-site</td>
</tr>
<tr class="row-even"><td>bsC</td>
<td>different-ratios</td>
<td>branch-site</td>
</tr>
<tr class="row-odd"><td>bsD</td>
<td>different-ratios</td>
<td>branch-site</td>
</tr>
<tr class="row-even"><td>b_free</td>
<td>positive-selection</td>
<td>branch</td>
</tr>
<tr class="row-odd"><td>b_neut</td>
<td>relaxation</td>
<td>branch</td>
</tr>
<tr class="row-even"><td>fb</td>
<td>free-ratios</td>
<td>branch</td>
</tr>
</tbody>
</table>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>model_name</strong> &#8211; a string like &#8220;model-name[.some-secondary-name]&#8221; (e.g.: &#8220;fb.my_first_try&#8221;, or just &#8220;fb&#8221;)
* model-name is compulsory, is the name of the model (see table above for the full list)
* the second part is accessory, it is to avoid over-writing models with the same name.</li>
<li><strong>ctrl_string</strong> &#8211; list of parameters that can be used as control file.</li>
<li><strong>kwargs</strong> &#8211; extra parameters should be one of: verbose, CodonFreq, ncatG, cleandata, fix_blength, NSsites, fix_omega, clock, seqfile, runmode, fix_kappa, fix_alpha, Small_Diff, method, Malpha, aaDist, RateAncestor, outfile, icode, alpha, seqtype, omega, getSE, noisy, Mgene, kappa, treefile, model, ndata.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="ete2.EvolNode.sep">
<code class="descname">sep</code><em class="property"> = '\n'</em><a class="headerlink" href="#ete2.EvolNode.sep" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ete2.EvolNode.show">
<code class="descname">show</code><span class="sig-paren">(</span><em>layout=None</em>, <em>tree_style=None</em>, <em>histfaces=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ete2.EvolNode.show" title="Permalink to this definition">¶</a></dt>
<dd><p>call super show of PhyloTree
histface should be a list of models to be displayes as histfaces</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>layout</strong> &#8211; a layout function</li>
<li><strong>tree_style</strong> (<em>None</em>) &#8211; tree_style object</li>
<li><strong>Nonehistface</strong> &#8211; an histogram face function. This is only to plot selective pressure among sites</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="ete2.EvolNode.write">
<code class="descname">write</code><span class="sig-paren">(</span><em>features=None</em>, <em>outfile=None</em>, <em>format=10</em><span class="sig-paren">)</span><a class="headerlink" href="#ete2.EvolNode.write" title="Permalink to this definition">¶</a></dt>
<dd><p>Inherits from Tree but adds the tenth format, that allows to display marks for CodeML.
TODO: internal writting format need to be something like 0</p>
<p>Returns the newick representation of current node. Several
arguments control the way in which extra data is shown for
every node:</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>features</strong> &#8211; a list of feature names to be exported
using the Extended Newick Format (i.e. features=[&#8220;name&#8221;,
&#8220;dist&#8221;]). Use an empty list to export all available features
in each node (features=[])</li>
<li><strong>outfile</strong> &#8211; writes the output to a given file</li>
<li><strong>format</strong> (<em>10</em>) &#8211; defines the newick standard used to encode the
tree. See tutorial for details.</li>
<li><strong>format_root_node</strong> (<em>False</em>) &#8211; If True, it allows features
and branch information from root node to be exported as a
part of the newick text string. For newick compatibility
reasons, this is False by default.</li>
<li><strong>is_leaf_fn</strong> &#8211; See <code class="xref py py-func docutils literal"><span class="pre">TreeNode.traverse()</span></code> for
documentation.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p><strong>Example:</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">t</span><span class="o">.</span><span class="n">get_newick</span><span class="p">(</span><span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s">&quot;species&quot;</span><span class="p">,</span><span class="s">&quot;name&quot;</span><span class="p">],</span> <span class="n">format</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="attribute">
<dt id="ete2.EvolTree">
<code class="descname">EvolTree</code><a class="headerlink" href="#ete2.EvolTree" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal"><span class="pre">EvolNode</span></code></p>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/ete_logo.png" alt="Logo"/>
            </a></p>
  <h4>Previous topic</h4>
  <p class="topless"><a href="reference_webplugin.html"
                        title="previous chapter">WebTreeApplication object</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_ncbi.html"
                        title="next chapter">NCBITaxa class</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/reference/reference_evoltree.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="reference_ncbi.html" title="NCBITaxa class"
             >next</a> |</li>
        <li class="right" >
          <a href="reference_webplugin.html" title="WebTreeApplication object"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >ETE&#8217;s Reference Guide</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Jaime Huerta-Cepas.
      Last updated on Nov 05, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>